[mlir][bufferization] Support custom types (1/N)

andrey-golubev opened the pull request:
Following the addition of TensorLike and BufferLike type interfaces (see 00eaff3e9c897c263a879416d0f151d7ca7eeaff), introduce minimal changes required to bufferize a custom tensor operation into a custom buffer operation.

To achieve this, new interface methods are added to TensorLike type interface that abstract away the differences between existing (tensor -> memref) and custom conversions.

The scope of the changes is intentionally limited (for example, BufferizableOpInterface is untouched) in order to first understand the basics and reach consensus design-wise.

---
Notable changes:
* mlir::bufferization::getBufferType() returns BufferLikeType (instead of BaseMemRefType)
* ToTensorOp / ToBufferOp operate on TensorLikeType / BufferLikeType. Operation argument "memref" renamed to "buffer"
* ToTensorOp's tensor type inferring builder is dropped (users now need to provide the tensor type explicitly)

andrey-golubev committed:
[mlir][bufferization] Support custom types (1/N)

Following the introduction of TensorLike and BufferLike type interfaces
(see 00eaff3e9c897c263a879416d0f151d7ca7eeaff), introduce minimal
changes required to bufferize a custom tensor operation into a custom
buffer operation.

To achieve this, a new conversion dialect interface is added that
abstracts away the differences between existing (tensor -> memref) and
custom conversions.

The scope of the changes is intentionally limited (for example,
BufferizableOpInterface is untouched) in order to first understand the
basics and reach consensus design-wise.

andrey-golubev commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizationConversionInterface.h:
```diff
@@ -0,0 +1,72 @@
+
+/// This class defines a virtual interface for conversions between tensor-like
+/// and buffer-like types.
+struct ConversionDialectInterface
```
@matthias-springer I ended up going with a separate interface (instead of putting this functionality into TensorLike / BufferLike type interfaces).

matthias-springer replied:
Why is this a dialect interface?

andrey-golubev replied:
the dialect interface is now gone!

hanhanW reviewed (commented):
Many thanks for the upstream contribution! I've seen some changes like this that'd break downstream projects. It is very reasonable because you're improving the bufferization.

I'd appreciate it if you can list down some naming changes in the PR description before you land the PR. It'd save us some time to look at all the details. E.g., the operand name of `ToTensorOp` becomes `buffer`; the return type of `bufferization::getBufferType` becomes `FailureOr<BufferLikeType>`, etc;

(It is definitely not a requirement, and it is okay if you don't do it. I appreciate your contribution. Thanks again! 🙏)

andrey-golubev commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td:
```diff
@@ -473,19 +482,19 @@ def Bufferization_ToTensorOp : Bufferization_Op<"to_tensor", [
-      auto rtt = memref::getTensorTypeFromMemRefType(memref.getType());
-      build($_builder, $_state, rtt, memref, restrict, writeable);
+    OpBuilder<(ins "Value":$buffer, CArg<"bool", "false">:$restrict, CArg<"bool", "false">:$writeable), [{
+      auto rtt = bufferization::detail::getTensorFromBuffer(buffer.getType());
```
@matthias-springer I took a look at the current implementation. I think there's a problem with this API if we switch to a BufferLike's function:
* ToTensor seems to infer the tensor type from buffer type
* Without options here, there's no (good) way to customize the (reverse) bufferization behavior for builtins

Should this be changed somehow? The things I could think of:
1. Drop type inference and make ToTensor always be constructed with an explicit type (so, user has to care about reverse bufferization)
2. Pass bufferization options to the builder
3. Assume that the current API is the way it is -> meaning that options cannot be used as customization point for builtins also and a different mechanism is needed
4. something else?

andrey-golubev replied:
> If we go with the above-mentioned verifyCompatibleTensorType approach, type inference must be dropped, indeed. I think that's alright. We can add some helper functions to make it easy for folks to migrate their code.

Turns out this builder *was* such a helper function. In fact, many places seem to just be able to provide the valid tensor type without additional reverse conversions.

andrey-golubev commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td:
```diff
@@ -387,20 +388,28 @@ def Bufferization_DeallocTensorOp : Bufferization_Op<"dealloc_tensor",
+class Bufferization_TensorAndBufferMatch<string tensor, string buffer> : PredOpTrait<
+  "specified tensor and buffer types match",
+  CPred<
+    "::mlir::bufferization::detail::typesMatchAfterBufferization("
```
@matthias-springer this would be the other problematic place. With the current approach, we want to validate that bufferization is "valid" on a tensor <-> buffer level.

The current logic checks `tensor.getShape() == buffer.getShape()` and `tensor.getElementType() == buffer.getElementType()` this practically means that TensorLike and BufferLike are ShapedType (fine by me), but even that is not enough. We've recently started to experiment with shape bounds and dynamic shapes -> getShape checking might not be sufficient.

Instead, I think we should either restore the old comparison logic (which was changed in https://github.com/llvm/llvm-project/commit/ced2fc7819d5ddea616ec330f18e08ff284c1868) or - more likely - have this put into an interface so that it's a customization point.

But then, which interface? TensorLike? BufferLike? Since it's a type matching function, it's kind of valid to be in both.

matthias-springer replied:
How about some kind of double dispatch? E.g., for `to_tensor`:

- Op verifier checks that operand type implements `BufferLikeTypeInterface`.
- Op verifier checks that result type implements `TensorLikeTypeInterface`.
- Op verifier calls `bufferLikeType.verifyCompatibleTensorType(tensorLikeType)`. The result is `std::optional<LogicalResult>`. `success` means that the type is compatible, `failure` means that it is incompatible, `nullopt` means that we don't know.
- In case of `nullopt`, call `tensorLikeType.verifyCompatibleBufferType(bufferLikeType)`.
- If the result is still `nullopt`, we fail verification because neither of the two types "know" each other.

andrey-golubev replied:
I wouldn't go with double dispatch to be honest because there's majorly no difference between "tensor equivalent to buffer" vs "buffer equivalent to tensor" (we have both things which do not change). For the time being, I guess we just put it somewhere? (either to buffer-like or to tensor-like). Perhaps with more changes it would be clearer what to do here.

andrey-golubev replied:
I see what you mean now. It's actually more (or less?) straightforward:
What we have is:
1. `custom.tensor` -> `custom.buffer`  <-- this is "problem-free" (because type interface implementation is on us)
2. `tensor<..., {custom encoding}>` -> `memref<..., {custom layout}>` <-- this is meh

there's never really a situation where we'd bufferize builtin into non-builtin or non-builtin into builtin, but the case is interesting (I'd perhaps add support for this separately if that has any use).

andrey-golubev replied:
Implemented the half of the querying for TensorLikeType.

andrey-golubev committed:
[NFC] Remove type-inferring builder of ToTensorOp

The builder is ambiguous given customizable tensor-like -> buffer-like
conversion and is thus removed. The places where reverse bufferization
has to happen rely on the pre-existing functionality.

andrey-golubev committed:
Switch from ConversionDialectInterface to TensorLike API

Noteworthy changes:
* bufferization::getMemRefType() accepts a TensorType instead of Value
  to achieve broader applicability
* BufferizationOptions::UnknownTypeConverterFn accepts a TensorType
  instead of Value to allow it being used in the updated getMemRefType()

matthias-springer reviewed (commented):
Looks good overall!

andrey-golubev commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizationTypeInterfaces.td:
```diff
@@ -21,10 +21,31 @@ def Bufferization_TensorLikeTypeInterface
+      /*methodName=*/"getBufferType",
+      /*args=*/(ins
+        "const ::mlir::bufferization::BufferizationOptions &":$options,
+        "const ::mlir::bufferization::BufferizationState &":$state,
```
honestly, i have no idea what this state thing is... I don't really need it myself, imho options is enough. I saw that `getBufferType` accepts it now, so there's a chance it's actually used by something? (no clue). let me try to make some sense out of it.

andrey-golubev commented on mlir/lib/Dialect/Bufferization/IR/BufferizationDialect.cpp:
```diff
@@ -57,7 +57,40 @@ struct BufferizationInlinerInterface : public DialectInlinerInterface {
+      const BufferizationState &state,
+      llvm::function_ref<mlir::InFlightDiagnostic()> emitError) const {
+    auto tensorType = cast<TensorType>(tensor);
+    // Fall back to tensor -> memref conversion.
```
never mind, outdated comment that i missed when moving stuff around.

andrey-golubev commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h:
```diff
@@ -267,7 +268,7 @@ struct BufferizationOptions {
   /// Parameters: Value, memory space, bufferization options
   using UnknownTypeConverterFn = std::function<BaseMemRefType(
-      Value, Attribute memorySpace, const BufferizationOptions &)>;
+      TensorType, Attribute memorySpace, const BufferizationOptions &)>;
```
I need this together with `getMemRefType()` API change since TensorLike::getBufferType() works with *types* (there are no values). see https://github.com/llvm/llvm-project/pull/142986/files#diff-93e08b1e03259ffa2ff33aec1bd8f907f066e1d86cf83cf140fb031fb31c6f3aR71-R72:
```cpp
llvm::FailureOr<BufferLikeType> getBufferType(mlir::Type tensor, // mlir::Type here now
  const BufferizationOptions &options,
  const BufferizationState &state,
  llvm::function_ref<mlir::InFlightDiagnostic()> emitError) const {
  auto tensorType = cast<TensorType>(type);

  // ...
      
  return cast<BufferLikeType>(
    getMemRefType(tensorType, options, /*layout=*/{}, *memSpace));
}
```

andrey-golubev replied:
I could separate this into another patch if you prefer

andrey-golubev commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td:
```diff
@@ -442,11 +451,11 @@ def Bufferization_ToTensorOp : Bufferization_Op<"to_tensor", [
+  let results = (outs Bufferization_TensorLikeTypeInterface:$result);
 
   let extraClassDeclaration = [{
     /// The result of a to_tensor is always a tensor.
```
self-review: `ToTensorOp::getType()` should return TensorLike now.

andrey-golubev committed:
Address code review feedback

* Remove BufferizationState from TensorLikeType::getBufferType()
* Rename castToMemRef to asMemRefType (+ add extra docs)
* Improve ToTensorOp's docs
* Apply minor suggestions

andrey-golubev committed:
Update ToTensorOp::getType() to return TensorLikeType

andrey-golubev commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h:
```diff
@@ -738,6 +740,14 @@ AliasingValueList unknownGetAliasingValues(OpOperand &opOperand);
 bool defaultHasTensorSemantics(Operation *op);
+
+/// This is a helper function used when buffer type is guaranteed to be memref.
+FailureOr<BaseMemRefType> castToMemRef(FailureOr<BufferLikeType> bufferType);
```
done.

hanhanW commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h:
```diff
@@ -739,6 +740,19 @@ AliasingValueList unknownGetAliasingValues(OpOperand &opOperand);
 /// BufferizableOpInterface::hasTensorSemantics
 bool defaultHasTensorSemantics(Operation *op);
+
+/// This is a helper function used when buffer type is guaranteed to be memref.
+/// It performs two actions: failure state checking and an explicit llvm::cast<>
+/// from the buffer-like type interface to a BaseMemRefType. This allows easier
+/// management of differences in C++ types at the API boundaries. Valid buffer
+/// type is casted to the memref type. Otherwise, the failure state is
+/// propagated i.e. asMemRefType(mlir::failure()) returns mlir::failure().
+FailureOr<BaseMemRefType> asMemRefType(FailureOr<BufferLikeType> bufferType);
```
It is weird that you declare this in `detail` namespace, but you use it like public API in other places. My understanding is that the methods within `detail` namespace are not intended to be used widely. With your change, it is used in many other dialects like Linalg/SCF/Arith/etc. Are you going to remove this function? Is it just for the transition state?

hanhanW replied:
I see, thanks for the explanation!

hanhanW commented on mlir/lib/Dialect/Bufferization/IR/BufferizableOpInterface.cpp:
```diff
@@ -1030,7 +1027,7 @@ bufferization::detail::unknownGetAliasingValues(OpOperand &opOperand) {
 
 bool bufferization::detail::defaultHasTensorSemantics(Operation *op) {
-  auto isaTensor = [](Type t) { return isa<TensorType>(t); };
+  auto isaTensor = [](Type t) { return isa<TensorLikeType>(t); };
```
Just a note, since I spent some time on this. Because the interface implementation for builtin types locate at `bufferization::BufferizationDialect::initialize()`, any pass that expects the registration need to add the bufferization to the dependent dialect.

> Dialects must be loaded in the MLIRContext before entities from these dialects (operations, types, attributes, …) can be created. 

https://mlir.llvm.org/docs/PassManagement/#dependent-dialects

My understanding was that a dialect should be registered if we're going to create the ops/types/attrs. The rule no longer hold today. 

https://github.com/llvm/llvm-project/blob/12409a18f609f965fbd40aba77a0b50dcdba147f/mlir/lib/Dialect/Bufferization/IR/BufferizationDialect.cpp#L102-L122

andrey-golubev replied:
Please refer to https://github.com/llvm/llvm-project/pull/134220 for more details.
I guess builtin dialect (and thus types, etc.) are supposed to be always loaded?

As for bufferization, I guess you would load bufferization dialect if you plan to use one-shot bufferization or similar facility. This would automatically register the builtin external models for you.

> any pass that expects the registration need to add the bufferization to the dependent dialect.

If your pass uses one-shot-bufferization, the Bufferization dialect must be a dependency anyway I guess?

> My understanding was that a dialect should be registered if we're going to create the ops/types/attrs. The rule no longer hold today.

Why do you think so? We've had some debate about how this part should look like for builtins in particular, the "let's put external models into dialect initialization" seemed to be a decent compromise given that builtin dialect cannot depend on bufferization.
